Feature: Implement #309 create jestrunner.configPath mapping

jeremygiberson opened the pull request:
Hi all. Scratching a personal itch that at least [a couple others seem interested in](https://github.com/firsttris/vscode-jest-runner/issues/309).

Adds the ability to specify a glob mapping for jest runner such as:
```json
{
  "jestrunner.configPath": {
    "**/*.it.spec.ts": "./jest.it.config.js",
    "**/*.spec.ts": "./jest.unit.config.js"
  }
}
```

- [x] updated contributes.configuration schema to make jestrunner.configPath be null, string or Record<string,string>
- [x] updated jestRunnerConfig getJestConfig to return path specified for matching glob
- [x] added unit test for configPath glob functionality
  - used a slightly tweaked approach to windows/linux grouping to reduce test boilerplate 
- [x] updated extension readme w/ a blurb about how to use glob map for configPath

Id like to point out that I think the [while loop](https://github.com/firsttris/vscode-jest-runner/blob/930a1776e5b5a77e3f6a6a9111eab7a94d07795b/src/jestRunnerConfig.ts#L107) in JestRunnerConfig's `findConfigPath` is not quite correctly implemented. When implementing my tests for the case that glob match was not found and projectPath was not specified, I **expected** to be able to assert against fs.existSync being called with a jest config in the workspace directory. Alas, the loop terminates before scanning the workspace directory. 

I don't know if this is intentional but it goes counter to my expectations. I made by test pass by asserting against the targetPath directory but I think the function should look in the workspace directory. Note, I'm new to vscode extension development and I'm assuming workspace directory = project root in most cases -- but maybe I'm making an incorrect assumption..

jeremygiberson commented:
Yea, that sounds correct. I can update the pr with that and add tests if you want. Unless you'd prefer to address it in a separate pr.

firsttris commented:
would be nice if you could include it in your PR and check if its still green.

firsttris commented:
this termination codition is actually in two places

currenPackagePath and findConfigPath

jeremygiberson commented:
I updated the code to traverse up to and include the workspace directory. 

I noticed a new issue during testing (windows os) in regards to how config paths are resolved:
 - path.resolve(path) results in `C:\\path\\to\\foo` 
 - jestRunnerConfig.findConfigPath returns the `path.join`'d path which fs.existsSync returns true for, presumably a `C:\\path\\to\\foo` format based on behavior of path.resolve.
 - jestRunnerConfig.getJestConfigPath returns 
     - if configPath is a string, the normalized (`\\ converted to /`) value of the specified string 
     - if configPath is a globMap, the normalized (`\\ converted to /`) value of the matching blog entry (I chose to match the behavior of string value)
     - if configPath is not defined or is a globMap and none of the globs match, returns result of jestRunnerConfig.findConfigPath (which doesn't call normalizePath)

I'm not sure why `normalizePath` exists. My guess is that maybe when we build the jest command jest would prefer a path with `/` rather than `\\`. If that's the case, I would think findConfigPath should also call normalizePath.

I didn't want to change behavior without fully understanding the reason behind the current behavior and whether it's intentional.

firsttris commented:
i resolved the conflicts with master, you can check if i did it correctly.

firsttris commented:
your right, jest needs unix path and this is why normalizePath exists. findConfigPath should also call normalizePath. where every we use the path its probably always better to have unix path.

jeremygiberson commented:
Alright, given that context I've updated my code to be consistent with the normalizePath behavior, and updated tests to reflect the expected behavior. 

I don't have any more reservations.

firsttris commented:
@jeremygiberson should i add you as collaborator to the repository?

jeremygiberson commented:
> @jeremygiberson i appreciate your contribution, should i add you as collaborator to the repository?

If it's just attribution for having contributed, sure. But if it's more a show of active engagement then I wasn't intending more contributions in the future unless something comes up that directly affects me.
